M2/Macaulay2/d/evaluate.d

```diff
@@ -1236,16 +1236,37 @@ tryEval(c:Code):Expr := (
 	else tryEvalSuccess = true);
     p);
 
+nullify(c:Code):Expr := (
+    e := tryEval(c);
+    if tryEvalSuccess
+    then (
+	when e
+	is Nothing do e
+	else (
+	    f := lookup(Class(e), QuestionQuestionS);
+	    if f == nullE then e
+	    else applyEE(f, e)))
+    else nullE);
+
+nullCoalescion(lhs:Code,rhs:Code):Expr := (
+    e := nullify(lhs);
+    when e
+    is Nothing do eval(rhs)
+    else e);
+setup(QuestionQuestionS, nullify, nullCoalescion);
+
 augmentedAssignmentFun(x:augmentedAssignmentCode):Expr := (
     when lookup(x.oper.word, augmentedAssignmentOperatorTable)
     is null do buildErrorPacket("unknown augmented assignment operator")
     is s:Symbol do (
 	-- evaluate the left-hand side first
 	lexpr := nullE;
-	if s.word.name === "??" -- null coalescion; ignore errors
+	if s.word.name === "??" -- x ??= y is treated like x ?? (x = y)
 	then (
-	    e := tryEval(x.lhs);
-	    if tryEvalSuccess then lexpr = e)
```

Comment:
I'm a little confused. Where is the actual assignment to `lexpr` happening now?

Reply:
We don't need it anymore when the left-hand side is non-null since we return right away.  And when it's null, we've already assigned `lexpr` as `nullE` a couple lines earlier.

Reply:
The assignment happens in one of several places, depending on what the code looks like.  

`lexpr` is the *evaluated* left-hand side.  But we still have the original unevaluated `Code` object, which we need to figure out what type of assignment we'll end up doing (global, local, assigning an element of a list/hash table, etc.).  Here's the basic gist:

* Go ahead and evaluate the left hand side (`lexpr`) and stuff it in an `evaluatedCode` object (`left`). We also check for errors here and whether someone has installed an augmented assignment method for this operator for the class of `lexpr`.  If so, call that and return the result.
* Now go back and look at the original unevaluated `Code` object to figure out how to assign things.  But instead of using this `Code` object, we swap it out with `left` to avoid re-evaluating the left-hand side.